### DISCUSSION TITLE
Does OPA have a REST API that accepts both JSON & REGO to make a decision?

### DISCUSSION THREAD

**Discussion by scaledapps** [open-policy-agent/community#289] [OPA and Rego]
Does OPA have a REST API that accepts both JSON & REGO to make a decision, instead of making calls to 2 separate APIs as follows?
1. curl -X PUT --data-binary @policy1.rego localhost:9090/v1/policies/play
2. curl -X POST localhost:9090/v1/data/play/allow -d @input.json -H 'Content-Type: application/json'

I need an API similar the following command:
`./opa eval -i input1.json -d policy1.rego`

**Comment by peteroneilljr** [comment]
Hi @scaledapps 

You should be able to use the query API to complete this.

https://www.openpolicyagent.org/docs/latest/rest-api/#execute-an-ad-hoc-query

If you can share some sample rego and data, that would help to generate the sample curl request.

**Reply by scaledapps** [reply]
Thanks @peteroneilljr.

Here's a simple Rego:
```
package play

default allow = false
allow = true {
    input.method == "GET"
    input.path = ["getSalary"]
}
```

And here's the input for it:
```
{
    "method": "GET",
    "path": ["getSalary"]
}
```
Can you please help with a CURL request for Query API that uses both?

**Reply by srenatus** [reply]
I'm not sure it helps. The Query API takes just that -- a query. And input. But it wouldn't let you push a rego file defining rules, and query them in the same step, I think.

I'm afraid there is no API similar to your CLI `opa eval` example. But it should be straightforward to build a little stateless "eval service" with just `github.com/open-policy-agent/opa/sdk` and some http server framework.

**Reply by anderseknert** [reply]
Just out of curiosity; if you have the policies and input data locally, what would doing this in the OPA server provide you that `opa eval` or `opa exec` doesn't?

**Reply by scaledapps** [reply]
Thanks @srenatus.

@anderseknert : here's my requirement...
Each API has unique policies associated with it.
Envoy Lua filter determines which policies are applicable to the current request, and asks OPA if the request should be allowed or refused.

![image](https://user-images.githubusercontent.com/14068511/196924096-fe3fdc60-34c5-4a37-a916-d66ba4d7999c.png)

**Reply by anderseknert** [reply]
Alright! I had not heard of using the Lua filter for authorization before. Would you be able to use something like [dynamic policy composition](https://www.styra.com/blog/dynamic-policy-composition-for-opa/) in OPA instead, to determine which policies that apply to a specific request?
